pkg/operator/csi/csidefaultstorageclasscontroller/csi_default_storageclass_controller.go

```diff
@@ -0,0 +1,131 @@
+// Based on the current StorageClasses in the cluster the controller can decide not
+// to deploy given StorageClass as default if there is already an existing default.
+// Also, the controller will emit alerts if more than one StorageClass defined
+// as default exists - this is not a recommended setup.
+// It produces following Conditions:
+// DefaultStorageClassControllerDegraded - failed to apply StorageClass provided
+type CSIDefaultStorageClassController struct {
+	name               string
+	assetFunc          resourceapply.AssetFunc
+	file               string
+	kubeClient         kubernetes.Interface
+	storageClassLister v1.StorageClassLister
+	operatorClient     v1helpers.OperatorClient
+	eventRecorder      events.Recorder
+}
+
+func NewCSIDefaultStorageClassController(
+	name string,
+	assetFunc resourceapply.AssetFunc,
+	file string,
+	kubeClient kubernetes.Interface,
+	InformerFactory informers.SharedInformerFactory,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &CSIDefaultStorageClassController{
+		name:               name,
+		assetFunc:          assetFunc,
+		file:               file,
+		kubeClient:         kubeClient,
+		storageClassLister: InformerFactory.Storage().V1().StorageClasses().Lister(),
+		operatorClient:     operatorClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().WithSync(
+		c.Sync,
+	).ResyncEvery(
+		time.Minute,
+	).WithSyncDegradedOnError(
+		operatorClient,
+	).WithInformers(
+		operatorClient.Informer(),
+		InformerFactory.Storage().V1().StorageClasses().Informer(),
+	).ToController(
+		"DefaultStorageClassController",
+		eventRecorder,
+	)
+}
+
+func (c *CSIDefaultStorageClassController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("DefaultStorageClassController sync started")
+	defer klog.V(4).Infof("DefaultStorageClassController sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	syncErr := c.syncStorageClass(ctx)
+
+	return syncErr
+}
+
+func (c *CSIDefaultStorageClassController) syncStorageClass(ctx context.Context) error {
+	expectedScBytes, err := c.assetFunc(c.file)
+	if err != nil {
+		return err
+	}
+
+	expectedSC := resourceread.ReadStorageClassV1OrDie(expectedScBytes)
+
+	existingSCs, err := c.storageClassLister.List(labels.Everything())
+	if err != nil {
+		klog.V(2).Infof("could not list StorageClass objects")
+		return err
+	}
+
+	defaultScCount := 0
+	for _, sc := range existingSCs {
+		if sc.Annotations[defaultScAnnotationKey] == "true" && sc.Name != expectedSC.Name {
+			defaultScCount++
+		}
+	}
+
+	if defaultScCount > 0 {
+		klog.V(2).Infof("default StorageClass already defined in cluster, %v StorageClass will be applied as non-default", expectedSC.Name)
+		expectedSC.Annotations["storageclass.kubernetes.io/is-default-class"] = "false"
+	}
```

Comment:
I did not test it in a real cluster, what happens when user does not want any default storage class in the cluster? From the code above it seems the controller will mark its own SC as default. It should do so only when *creating* the SC, it should not overwrite the annotation when user set it manually to `false`.

Reply:
I just tested it, and yeah I did not realize we need to differentiate creation from re-apply. I've changed the code a bit to fix this and added a unit test for this case. Good catch.